### PR TITLE
Coupon expiry

### LINKED ISSUE #348: Add coupon expiry and create new one

### PR DESCRIPTION
**Pull request by ganhongyao**
Closes #348 

image

image

### CODE REVIEW

File: backend/app/models/coupon.rb [hunk 1]
```diff
@@ -4,19 +4,35 @@ class Coupon < ApplicationRecord
...
-  validates_associated :secondary_donation
+  validate :coupon_expires_before_campaign_ends
 
   def redeemed?
-    campaign_charity_id.present?
+    redemption_id.present?
+  end
+
+  def expired?
+    expires_at < DateTime.current
```
Comment (sivayogasubramanian) [inline backend/app/models/coupon.rb:22]: When creating we use Date.current and here we use DateTime.current. Is that okay?

File: backend/config/routes.rb [hunk 2]
```diff
@@ -39,6 +40,8 @@
 
       resources :primary_donors
 
+      resources :redemptions, only: %i[create]
```
Comment (sivayogasubramanian) [inline backend/config/routes.rb:43]: This is never used right since we expose the route in coupons.

**Review by sivayogasubramanian — APPROVED** [review]
LGTM! Only checked briefly to unblock others.

**Comment by sivayogasubramanian** [comment]
The seeds also need to be updated and there are some merge conflicts.

File: backend/config/routes.rb (hunk 2 above)
Comment (ganhongyao) [reply backend/config/routes.rb:42]: Removed the route in coupons in https://github.com/Giving-Coupons/giving-coupons/pull/373/commits/69976800aa6060b035542a0b329e6d787bebf4c3

**Pull request merged**